File: bishop/src/bishop/base/LineAttackTable.java
package bishop.base;

import bishop.tables.BetweenTable;

/**
 * This table contains squares attacked from some square with some occupancy.
 * @author Ing. Petr Ležák
 */
public class LineAttackTable {
	
	private static long[] attackTable = initializeTable(0);
	private static long[] pinTable = initializeTable(1);
	
	/**
	 * Returns mask of attacked squares.
	 * @param index line index
	 * @return mask of attacked squares
	 */
	public static long getAttackMask (final int index) {
		return attackTable[index];
	}

	/**
	 * Returns mask of squares blocked by one piece.
	 * @param index line index
	 * @return mask of attacked squares with one blocking piece 
	 */
	public static long getPinMask (final int index) {
		return pinTable[index];
	}

	public static long getLineMask (final int square, final FileRankOffset offset) {
	    long mask = 0;

    	int targetFile = Square.getFile(square) + offset.getFileOffset();
    	int targetRank = Square.getRank(square) + offset.getRankOffset();

    	while (File.isValid(targetFile) && Rank.isValid(targetRank)) {
    		final int targetSquare = Square.onFileRank(targetFile, targetRank);
    		mask |= BitBoard.getSquareMask(targetSquare);

    		targetFile += offset.getFileOffset();
    		targetRank += offset.getRankOffset();
    	}
    	
    	return mask;
	}
		
	private static long[] initializeTable(final int betweenCount) {
		final long[] table = new long[LineIndexer.getLastIndex()];
		
		for (int direction = CrossDirection.FIRST; direction < CrossDirection.LAST; direction++) {
			for (int square = Square.FIRST; square < Square.LAST; square++) {
				long mask = BitBoard.EMPTY;
				
				for (FileRankOffset offset: LineIndexer.getDirectionOffsets(direction)) {
					mask |= LineAttackTable.getLineMask(square, offset);
				}
				
				for (BitBoardCombinator combinator = new BitBoardCombinator(mask); combinator.hasNextCombination(); ) {
					final long occupancy = combinator.getNextCombination();
					final long attackMask = getAttackMask(square, mask, occupancy, betweenCount);
					
					final int index = LineIndexer.getLineIndex(direction, square, occupancy);
					table[index] = attackMask;
				}
			}
		}
		
		return table;
	}

	private static long getAttackMask (final int square, final long cross, final long occupancy, final int betweenCount) {
		long attackMask = BitBoard.EMPTY;
		
		for (BitLoop loop = new BitLoop(cross); loop.hasNextSquare(); ) {
			final int targetSquare = loop.getNextSquare();
			final long betweenOccupiedSquares = BetweenTable.getItem(square, targetSquare) & occupancy;
			
			if (BitBoard.getSquareCount(betweenOccupiedSquares) == betweenCount)
				attackMask |= BitBoard.getSquareMask(targetSquare);
		}
		
		return attackMask;
	}

}

File: native/bishop/base/line_attack_table.cpp


#include "line_attack_table.h"
#include "bit_board_combinator.h"
#include "bit_loop.h"
#include "between_table.h"


using namespace bishop::base;


const LineAttackTable::TheTable bishop::base::LineAttackTable::getAttackMask([](auto &table) { initializeTable (table, 0); });
const LineAttackTable::TheTable bishop::base::LineAttackTable::getPinMask([](auto &table) { initializeTable (table, 1); });


void bishop::base::LineAttackTable::initializeTable (TheTable &table, const int betweenCount) {
	for (CrossDirection::Type direction = CrossDirection::FIRST; direction < CrossDirection::LAST; direction++) {
		for (Square::Type square = Square::FIRST; square < Square::LAST; square++) {
			BitBoard::Type mask = BitBoard::EMPTY;
				
			for (FileRankOffset offset: LineIndexer::DIRECTION_OFFSETS[direction]) {
				mask |= LineAttackTable::getLineMask(square, offset);
			}
				
			for (BitBoardCombinator combinator(mask); combinator.hasNextCombination(); ) {
				const BitBoard::Type occupancy = combinator.getNextCombination();
				const BitBoard::Type attackMask = calculateAttackMask(square, mask, occupancy, betweenCount);
					
				const LineIndexer::IndexType index = LineIndexer::getLineIndex(direction, square, occupancy);
				table(index) = attackMask;
			}
		}
	}
}

BitBoard::Type bishop::base::LineAttackTable::getLineMask (const Square::Type square, const FileRankOffset offset) {
	BitBoard::Type mask = BitBoard::EMPTY;

	File::Type targetFile = Square::getFile(square) + offset.fileOffset;
	Rank::Type targetRank = Square::getRank(square) + offset.rankOffset;

    	while (File::isValid(targetFile) && Rank::isValid(targetRank)) {
		const Square::Type targetSquare = Square::onFileRank(targetFile, targetRank);
    		mask |= BitBoard::getSquareMask(targetSquare);

    		targetFile += offset.fileOffset;
    		targetRank += offset.rankOffset;
    	}
    	
    	return mask;
}


BitBoard::Type bishop::base::LineAttackTable::calculateAttackMask (const Square::Type square, const BitBoard::Type cross, const BitBoard::Type occupancy, const int betweenCount) {
	BitBoard::Type attackMask = BitBoard::EMPTY;
		
	for (BitLoop loop(cross); loop.hasNextSquare(); ) {
		const Square::Type targetSquare = loop.getNextSquare();
		const BitBoard::Type betweenOccupiedSquares = BetweenTable::getItem(square, targetSquare) & occupancy;
			
		if (BitBoard::getSquareCount(betweenOccupiedSquares) == betweenCount)
			attackMask |= BitBoard::getSquareMask(targetSquare);
	}
	
	return attackMask;
}

